### PR TITLE
[Draft] Add audited shared-expert overlap and Issue #28 campaign gates

### LINKED ISSUE #21: [Roadmap][Single-NPU] 扩展 LatchMoE 以支持 shared experts 与复杂路由
## Reviewer-facing objective

把 LatchMoE 从“只验证 Qwen3-30B-A3B 的单 routed-output 路径”扩展为一个按架构能力定义、可在单张 Ascend NPU 上支持 shared experts 与复杂路由的 graph-compatible MoE offloading runtime。

预期可辩护的系统结论是：当动态 routed experts 与 always-active shared computation 共存时，LatchMoE 仍能通过稳定 routed-slot storage、显式输出契约和路由语义保持 PIECEWISE graph replay、容量有界执行和原生模型语义。

### In scope

- 单张物理 Ascend NPU；
- BF16、未量化 MoE；
- PIECEWISE ACLGraph；
- routed-only、external shared expert、gated shared expert；
- external/internal router、grouped top-k、softmax/sigmoid、correction bias、routed scaling，以及可显式描述的复杂路由；
- Prefill、Decode 和 multi-wave overflow；
- native-vs-LatchMoE 的 router、layer、token、graph 和性能证据；
- fused/mix-placement shared expert 与 shared/H2D overlap 作为 stretch goals，在 backend 与 checkpoint 可运行时纳入结论。

### Explicitly out of scope

- TP/EP/DP/PCP 等多卡语义；
- W8A8、W4A8、FP8 等量化权重；
- prefix cache 与多请求并发；
- 为每个模型写不可复用的 model-name patch；
- 通过 eager fallback 获得“支持”或性能结果。

这些维度可在后续 roadmap 处理，不阻塞本轮 shared/router 架构扩展。

## Architecture coverage, not model counting

Core paper gate 至少验证以下三个 capability tuples；多个同构模型不算新的覆盖：

1. **Routed-only baseline**：Qwen3-30B-A3B，作为当前已验证对照。
2. **External shared + complex router**：优先使用 GLM-4.7-Flash，覆盖 external resident shared、sigmoid、correction bias 与 routed scaling。
3. **Gated shared + internal router**：优先使用 Qwen3-Next-80B-A3B-Instruct，覆盖 gated shared output、internal router 与更大的 routed expert pool。
4. **Fused/mix-placement path（stretch）**：优先复用 Qwen3-Next checkpoint，在 `mix_placement=false/true` 下分别验证 external/fused capability tuple；只有真实 NPU cell 通过后才进入论文 claim。

Qwen2-57B-A14B-Instruct 作为 gated external shared 的 fallback/large-shared-lane stress checkpoint。DeepSeek-V2-Lite 在 pinned Ascend 栈的 PIECEWISE graph preflight 通过前不进入主矩阵。

## Global correctness contract

- shared experts 默认是 always-active resident lane；external shared weights 与 shared gate 不参与 routed cache replacement。
- routed top-k ID/weight、shared gate、routed scaling、correction bias 和最终 combine 与 native 路径语义一致。
- multi-wave 只拆 routed token-expert pairs；shared contribution 每层、每 token 只生成并合并一次。
- custom-op 返回类型必须使用固定 schema 区分 single tensor 与 tuple；不能让同一 op 动态改变返回类型，也不能把 shared+routed 提前相加后绕过原生 reduce/scale contract。
- CPU-first loading、original-weight release、slot sizing 与 HBM ledger 必须区分 routed-offloaded weights 和 shared-resident weights。
- 不支持的能力组合在启动前结构化 fail closed。
- 任一 graph/eager fallback、silent route change、shared contribution 丢失或重复累加都算失败。

## Child issues and critical path

- [x] #27 Stage -1 / Stage 5 / P0：模型盘点、native graph preflight 与 qualification matrix v2
- [x] #22 Stage 0 / P0：能力契约与 fail-closed 支持矩阵
- [x] #23 Stage 1 / P0：external resident shared expert tuple seam 与 weight lifecycle
- [x] #24 Stage 2 / P0：复杂 router 语义等价
- [x] #25 Stage 3 / P1 stretch：fused/mix-placement fixed shared lane
- [x] #26 Stage 4 / P1 stretch：shared compute 与 routed staging/MLP overlap
- [x] #28 Stage 6 / P0：matched campaign 与论文 claim-evidence 闭环

依赖关系：

```text
#27 Phase A: config inventory + native PIECEWISE preflight
  --> #22 capability contract
        |-- #23 external shared + weight lifecycle --|
        |-- #24 complex routing ---------------------|--> #27 Phase B qualification --> #28 core campaign/paper
        |          |-- #25 fused shared (stretch) ---|
        |-- #26 overlap (stretch; external path after #23, fused path after #25)
```

#23 与 #24 可并行。#27 Phase A 必须先于 capability contract，避免先定义抽象再发现 checkpoint/backend 形态不一致；Phase B 在实现稳定后生成 registry 与 qualification matrix。

## Stage gates

- **G-1 Model preflight**：冻结 config/checkpoint digest、能力 tuple、权重/HBM 估算；native BF16 TP1 至少完成 loader/model-construction preflight，PIECEWISE blocker 必须显式记录。
- **G0 Capability contract**：结构化识别并报告 output/shared/router/combine/weight-lifecycle 能力；未知组合 fail closed。
- **G1 External shared correctness**：独立 tuple op、常驻 shared lane、模型通用内存账本和 multi-wave once-only combine 通过。
- **G2 Complex routing parity**：top-k IDs/weights、layer output 和 token output 对齐 native。
- **G3 Core multi-model qualification**：routed-only、external-complex 与 gated-internal 三个 capability tuples 完成 PIECEWISE qualification。
- **G4 Fused/mix-placement correctness（stretch）**：固定 shared lane 与动态 routed slots 的 ID、mapping、wave coverage 正确。
- **G5 Overlap and performance（stretch）**：shared compute、H2D staging 与 routed MLP 的 overlap 有时间线证据，且没有通过同步化换取正确性。

## Paper gate versus stretch gate

Core paper completion requires #22、#23、#24、#27 和 #28；#25/#26 只有在真实 NPU qualification 通过后才扩展 fused/overlap claims。#25 或 #26 的 backend blocker 不得阻塞 external shared + complex router 的较窄、可辩护结论。

## Relationship to existing work

- #4 继续负责 LatchMoE 与外部 placement policy 的边界。
- #1 继续负责插件生命周期和 pinned hook compatibility。
- #7/#11/#13/#17/#19 已完成的单模型 graph、异步生命周期、multi-wave 和安装证据继续作为 routed-only baseline，不重复执行其研究任务。

## Completion gate

Core child issues 关闭并同步更新 `paper/CLAIMS.md`、`paper/EVIDENCE_MAP.md`、Evaluation、Discussion 和匿名 artifact。代码合入但缺少 NPU qualification 不视为完成。Fused/overlap 仅在 #25/#26 完成时写入摘要、Introduction 和 Conclusion。

### LINKED ISSUE #28: [Stage 6][P0] 执行单卡多架构 matched campaign 并闭合论文 claim-evidence
Parent: #21  
Blocked by: #23, #24, #27  
Includes fused/overlap claims only if: #25, #26 are complete

## Research questions

- RQ1: LatchMoE 能否在 routed-only、external shared 和 complex-routing 架构上保持原生语义与 PIECEWISE graph replay？
- RQ2: 在相同 HBM budget 下，LatchMoE 相比 full residency/capacity failure、native prefetch、legacy layered 和 graph-compatible full-layer staging 的端到端表现如何？
- RQ3: fixed slots、multi-wave 和 shared/H2D overlap 各自贡献多少？
- RQ4: 结果对模型架构、slot capacity、prompt shape 和 route active-set 有多敏感？
- RQ5: shared resident memory、host memory、transfer bytes 和 graph/control overhead 的代价是什么？

## Fixed experimental scope

- 单张同型号 Ascend NPU；
- BF16 unquantized；
- TP1、max_num_seqs=1、client concurrency=1；
- prefix cache disabled；
- PIECEWISE graph only for headline comparisons；
- 相同 request manifest、启动顺序和 memory/KV contract。

## Required baselines

对每个可运行模型说明并执行最强可信基线：

- full-resident/no-offload；OOM 或容量失败作为结果保留；
- native vLLM/vLLM-Ascend expert prefetch；
- legacy layered path；
- LatchMoE full-layer staging；
- LatchMoE multi-wave；
- shared no-overlap/overlap ablation，仅在 #26 完成时。

不能运行的 baseline 必须给出可复核 blocker，不能静默删除。

## Workloads and sensitivity

每个 architecture tuple 至少运行：

- one-request correctness trigger；
- mixed-chat；
- prefill-heavy；
- decode-heavy；
- 至少三个可行 slot/capacity 点，包括一个压力或失败点。

每个有效性能单元至少三次独立服务启动，固定 AB/BA 顺序或等价 matched design。

## Metrics

- TTFT p50/p95、TPOT p50/p95、throughput 和完整 request distribution；
- HBM peak、KV headroom、host bytes、resident shared bytes、slot/stage bytes；
- H2D/D2D bytes、wait time、cache hit/miss、wave count；
- graph capture/replay、fallback、NPU/ACL/OOM、lease/mapping failures；
- overlap timeline 与 launch/control overhead。

报告 repeat-level dispersion；若进行显著性检验，同时报告 effect size 和 confidence interval。

## Claim gates

- **Generality claim**：至少三个不同 capability tuples 全部门禁通过。
- **Fused shared claim**：#25 的真实 NPU cell 通过；否则明确排除。
- **Overlap claim**：#26 matched ablation 和时间线同时通过。
- **Performance claim**：不能由跨模型或非 matched 历史数据支持。

## Paper and artifact updates

- 更新 `paper/CLAIMS.md` 与 `paper/EVIDENCE_MAP.md`；
- Motivation 加入跨架构 active-set/shared-memory characterization；
- Design 明确 shared lane、router contract 和 alternative choices；
- Evaluation 按 RQ 组织，不按模型逐个叙述；
- Discussion 保留单卡、BF16、无并发/无 prefix cache 边界；
- 打包匿名 raw manifests、logs、summaries、digests 和失败记录。

## Acceptance

所有写入摘要、Introduction 和 Conclusion 的数字均可追溯到 checked artifact；最终 paper 不再声称只支持单 routed-output，但也不越过本 issue 的单卡 BF16 证据边界。

### PR DESCRIPTION
**Pull request by Li-changwu**
## 概要

本 PR 建立并固化 LatchMoE shared-expert/offload 的实现、可审计实验工具和当前论文范围内的 claim-evidence 闭环。

- #26：resident shared expert 隐藏 routed expert H2D，已完成并关闭。
- #28：GLM 冻结 matched campaign 已通过 verifier；Qwen3-Next TP1 官方 GDN + LatchMoE offload correctness matched 已完成，#28 已按当前论文范围关闭。
- #21：本 PR 完成当前论文范围内的 shared/router/offload 证据闭环，合并后关闭。

## #26 结果

- locked latchmoe 环境、单 Ascend NPU、PIECEWISE graph；
- matched treatment/control 输出 byte-identical；
- raw hardware shared/H2D overlap：975.75 us，control：0 us；
- 182/182 个 H2D stage 安装 consumer dependency；
- serving 区间 device-wide sync：0；
- verifier 10/10 checks passed，host regression 95 passed。

### GLM matched campaign

冻结合同 issue28-glm-prefill-v1 在单卡、BF16、TP1、max_num_seqs=1、PIECEWISE 和固定 manifest/order/memory/KV 条件下完成：

- 15/15 units 执行完成；
- full-resident 三次容量失败按合同接受；
- native-prefetch、legacy-layered、LatchMoE full-layer、LatchMoE multi-wave 各三次成功；
- verifier passed，所有成功 arm 输出 token 一致。

该证据支持 GLM external-shared + complex-router 的窄范围 correctness/matched 结论，不扩展为广泛性能 superiority。

### Qwen3-Next official GDN matched correctness

在 latchmoe 环境中，TP1、num_slots=128、CPU-first、释放原始 expert weights、graph-compatible offload 下，未启用 PyTorch GDN fallback，官方 GDN 路径成功生成：

A gated delta network

48 个 MoE 层共完成 192 次 router parity 和 192 次 layer-boundary parity；top-k IDs、weights、routed/shared outputs 全部一致，最大绝对误差为 0。该结果证明当前 Qwen3-Next 可以在 LatchMoE offload 中采用官方 GDN。

更大范围的三架构、decode/mixed-chat 和 pressure-point 矩阵作为后续扩展，不阻塞当前论文写作。

## 证据边界

论文仍保留单卡、BF16、TP1、max_num_seqs=1、PIECEWISE、无 prefix cache 等边界，不把当前结果扩展为跨设备、并发或普遍性能 superiority claim。

实现、测试和证据说明见：
- docs/evidence/issue-28-glm-matched.md
- docs/evidence/issue-28-qwen3-next-qualification.md

Closes #28
Closes #21
Refs #26

### CODE REVIEW

**Comment by Li-changwu** [comment]
## Runtime installation update (2026-08-20)

The repository installer has now been executed successfully in the new Conda environment `latchmoe`; `vllm-hust-dev` was not modified.

The locked stack is:
- vLLM `0.21.0+empty`, commit `ad7125a431e176d4161099480a66f0169609a690`;
- hook seam `4806367eeeb7d62b32078ae90cd929cc06d825fe`;
- Torch/Torch-NPU `2.10.0/2.10.0.post2`;
- CANN `9.0.1`;
- `python -m vllm_moe_offload_ascend check --strict`: PASS.

The seam was installed with `COMPILE_CUSTOM_KERNELS=0), following the repository installation method and avoiding the unrelated full Ascend custom-kernel build. The previous r3 run failed before model execution because that full-build path left `_C_ascend.moe_gating_top_k` unavailable; it is invalid evidence and remains excluded.

Next: rerun #26 treatment/control with raw trace verification, then execute the frozen #28 matched campaign. The PR remains draft and no issue is being closed until the corresponding evidence gates pass.

**Comment by Li-changwu** [comment]
## 阻塞进一步收敛（2026-08-20）

补充一个重要边界：仓库 installer 使用 COMPILE_CUSTOM_KERNELS=0 是正确的 LatchMoE hook 安装方式，且 check --strict 已通过；但这不代表所有 vLLM-Ascend 模型路径都能在没有 _C_ascend 扩展时直接运行。

对当前 GLM-4.7-Flash，默认 device adaptor 会调用 _C_ascend.moe_gating_top_k。该扩展在 COMPILE_CUSTOM_KERNELS=0 下不存在，因此不能把下一次 GLM campaign 直接标记为可运行。已在 latchmoe + NPU5 + CANN 9.0.1 上验证 torch_npu.npu_moe_gating_top_k 对 GLM 的 64 experts/top-k=4 形状可用并通过小规模 NPU smoke。

因此当前真正的下一步是：在 LatchMoE runtime 中增加并测试 CANN9/无自定义扩展时的 native router fallback，或构建并验证最小必要算子；完成后才重新跑 #26 treatment/control 和 #28 campaign。此前 r3 仍是无效证据，所有 issue 保持 open。

**Comment by Li-changwu** [comment]
## 最新阶段结果（2026-08-20）\n\nr11 matched treatment/control 已完成并通过正确性与 H2D 依赖门禁，但 overlap 关闭门禁仍失败：原始 shared/routed=0 us、shared/H2D=0 us；treatment 还有一处 serving-path 。因此 PR 继续保持 draft，不使用虚假的 。\n\n当前已完成 CANN9 native router/init-routing fallback，并在  环境通过 host tests；下一步执行 #28 的冻结 matched campaign。

**Comment by Li-changwu** [comment]
补充更正：上一条评论中的若干代码标记因 shell 转义被省略，数据结论不变。

r11 matched treatment/control 已通过正确性与 H2D 依赖门禁，但原始 shared/routed=0 us、shared/H2D=0 us；treatment 还有一处 serving-path aclrtSynchronizeDeviceWithTimeout。因此 PR 继续保持 draft，不使用 Closes #26。CANN9 native router/init-routing fallback 已在 latchmoe 环境通过 host tests，下一步执行 #28 冻结 matched campaign。

**Comment by Li-changwu** [comment]
按协作安排暂缓：另一进程正在处理 #26，我已停止当前 #28 r5 campaign 并回收精确的 8106 server。已完成 artifacts 保留在 /workspace/latchmoe-issue28-campaign-r5，但未将未完成单元计入结果，也未把 PR 从 draft 改为 ready；待 #26 关闭后继续 matched campaign 和最终 verifier。

**Comment by Li-changwu** [comment]
阶段更新（commit cfc9d47）

- #28 GLM r5 matched campaign：15/15 units，verifier passed；
- Qwen3-Next qualification：在显式 `LATCHMOE_ENABLE_GDN_PYTORCH_FALLBACK=1` 下 status=ok；
- 新增 GLM/Qwen evidence ledger，并更新 `paper/CLAIMS.md` 与 `paper/EVIDENCE_MAP.md`；
- latchmoe 环境回归：119 passed；
- #21/#28 仍保持 open，PR 继续保持 draft，等待 native matched reference 和完整架构矩阵。
